src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/DiagnosticOr.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime.InteropServices.ComTypes;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.Interop
+{
+    /// <summary>
+    /// A union of Diagnostic and T, representing either a value of type T or a Diagnostic.
+    /// Useful helper functions exist in <see cref="IncrementalValuesProviderExtensions"/> and <see cref="IncrementalGeneratorInitializationContextExtensions"/>.
+    /// </summary>
+    public abstract record DiagnosticOr<T>
+    {
+        public abstract bool IsValue { get; }
+
+        public bool IsDiagnostic => !IsValue;
+
+        /// <summary>
+        /// Throws <see cref="InvalidOperationException"/> if IsValue is false
+        /// </summary>
+        public abstract T Value { get; }
+
+        /// <summary>
+        /// Throws <see cref="InvalidOperationException"/> if IsDiagnostic is false
+        /// </summary>
+        public abstract Diagnostic Diagnostic { get; }
```

Comment:
What do we think about creating a `DiagnosticTemplate` type here instead of passing a Diagnostic instance around? As you shared out to the team, there's issues around diagnostic equality (and after giving it a second look, your thought that the issue would affect us is correct).

Reply:
Yeah, that's a good idea, but maybe in another PR?

Reply:
Sure, we can do that as a follow up!